File: java/917.reverse-only-letters.java
/*
 * @lc app=leetcode id=917 lang=java
 *
 * [917] Reverse Only Letters
 */

// @lc code=start
class Solution {
    public String reverseOnlyLetters(String s) {
        char[] s2 = new char[s.length()];
        int left = 0, right = s.length() - 1;
        while (left <= right) {
            if (!Character.isAlphabetic(s.charAt(left))) {
                s2[left] = s.charAt(left);
                ++left;
            } else if (!Character.isAlphabetic(s.charAt(right))) {
                s2[right] = s.charAt(right);
                --right;
            } else {
                s2[left] = s.charAt(right);
                s2[right] = s.charAt(left);
                ++left;
                --right;
            }
        }
        return String.valueOf(s2);
    }
}
// @lc code=end

File: py/917.reverse-only-letters.py
#
# @lc app=leetcode id=917 lang=python3
#
# [917] Reverse Only Letters
#

# @lc code=start
class Solution:
    def reverseOnlyLetters(self, s: str) -> str:
        s2 = list(s)
        left = 0
        right = len(s)-1
        while left < right:
            if not s[left].isalpha():
                left += 1
            elif not s[right].isalpha():
                right -= 1
            else:
                s2[left], s2[right] = s2[right], s2[left]
                left += 1
                right -= 1
        return "".join(s2)

# @lc code=end
